imports/wasi_snapshot_preview1/poll.go

```diff
@@ -65,95 +77,148 @@ func pollOneoffFn(ctx context.Context, mod api.Module, params []uint64) syscall.
 	}
 
 	// Eagerly write the number of events which will equal subscriptions unless
-	// there's a fault in parsing (not processing).
+	// there'e a fault in parsing (not processing).
 	if !mod.Memory().WriteUint32Le(resultNevents, nsubscriptions) {
 		return syscall.EFAULT
 	}
 
 	// Loop through all subscriptions and write their output.
 
+	var stdinSubs []stdinEvent
+	var timeout time.Duration = 1<<63 - 1 // max timeout
+	readySubs := 0
+
 	// Layout is subscription_u: Union
 	// https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#subscription_u
 	for i := uint32(0); i < nsubscriptions; i++ {
 		inOffset := i * 48
 		outOffset := i * 32
 
 		eventType := inBuf[inOffset+8] // +8 past userdata
-		var errno syscall.Errno        // errno for this specific event (1-byte)
+		// +8 past userdata +8 contents_offset
+		argBuf := inBuf[inOffset+8+8:]
+		userData := inBuf[inOffset : inOffset+8]
+
+		evt := &event{
+			eventType: eventType,
+			userData:  userData,
+			errno:     wasip1.ErrnoSuccess,
+			outOffset: outOffset,
+		}
+
 		switch eventType {
 		case wasip1.EventTypeClock: // handle later
-			// +8 past userdata +8 contents_offset
-			errno = processClockEvent(ctx, mod, inBuf[inOffset+8+8:])
+			newTimeout, err := processClockEvent(argBuf)
+			if err != 0 {
+				return err
+			}
+			if newTimeout < timeout {
+				timeout = newTimeout
+			}
+			writeEvent(outBuf, evt)
 		case wasip1.EventTypeFdRead, wasip1.EventTypeFdWrite:
-			// +8 past userdata +8 contents_offset
-			errno = processFDEvent(mod, eventType, inBuf[inOffset+8+8:])
+			fsc := mod.(*wasm.ModuleInstance).Sys.FS()
+			fd := le.Uint32(argBuf)
+
+			stdinReader := processFDEvent(fsc, fd, evt)
+			if stdinReader == nil {
+				// if stdinReader is not an interactive session, then write back immediately
+				readySubs++
+				writeEvent(outBuf, evt)
+			} else {
+				// otherwise, delay processing with the timeout
+				stdinSubs = append(stdinSubs, stdinEvent{evt, stdinReader})
+			}
 		default:
 			return syscall.EINVAL
 		}
+	}
 
-		// Write the event corresponding to the processed subscription.
-		// https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#-event-struct
-		copy(outBuf, inBuf[inOffset:inOffset+8]) // userdata
-		if errno != 0 {
-			outBuf[outOffset+8] = byte(wasip1.ToErrno(errno)) // uint16, but safe as < 255
-		} else { // special case ass ErrnoSuccess is zero
-			outBuf[outOffset+8] = 0
+	// process timeout and interactive inputs (if any)
+	if timeout > 0 && readySubs == 0 {
```

Comment:
It would be great if it checks stdin even when timeout=0.

c.f. https://linux.die.net/man/3/poll

> If none of the defined events have occurred on any selected file descriptor, poll() shall wait at least timeout milliseconds for an event to occur on any of the selected file descriptors. If the value of timeout is 0, poll() shall return immediately. If the value of timeout is -1, poll() shall block until a requested event occurs or until the call is interrupted.

Reply:
Thanks for the comment.

> if stdin is interactive, then poll 0 would mean that the end user has not the chance to write anything interactively, so we may give up anyway

I think there are some cases where there are chances for user to write strings

e.g.

```c
#include <stdio.h>
#include <unistd.h>
#include <sys/select.h>

int main(int argc, char **argv)
{
  char buf[5];
  int len = 5, ret;
  fd_set rfds;
  struct timeval tv;

  for (int i = 0; i < 10; i++) {
    printf("checking inputs...\n");
    fflush(stdout);
    FD_ZERO(&rfds);
    FD_SET(0, &rfds);
    tv.tv_sec = 0;
    tv.tv_usec = 0;
    ret = select(1, &rfds, NULL, NULL, &tv);
    if ((ret > 0) && FD_ISSET(0, &rfds)) {
      ret = read(0, buf, len);
      if (ret > 0) {
        printf("got %.*s (first 5 chars)\n", ret, buf);
      }
    } else {
      printf("no input: %d\n", ret);
    }
    sleep(1); // do something
  }
}
```

host:

```console
$ ./a.out
checking inputs...
no input: 0
checking inputs...
no input: 0
hi
checking inputs...
got hi
 (first 5 chars)
```

But the current implementation of poll_oneoff can't receive the input so the above code doesn't work (i.e. input isn't detected).

Reply:
ah, great point. in fact, if I remove the `timeout > 0` (and TBF even if it's more expensive, the entire if check can be removed) the whole thing hangs as soon as there is one valid input 🤦 

```
❯ wazero run poll.wasm
checking inputs...
no input: 0
checking inputs...
no input: 0
x
checking inputs...
got x
 (first 5 chars)
checking inputs...
<hangs here>
```

Reply:
Maybe https://github.com/tetratelabs/wazero/pull/1346#discussion_r1162492497  is related?